Add lesson for Loops in Javascript (#101)
Ada-Developers-Academy/core-unit-3: learning-another-language/js-loops.md

	console.log(arr[i]);
}

// Output:
// foo
// bar
// baz
// undefined

// In the above example, arr.length = 3, 
// so accessing arr[3] leads to an undefined element
```

When iterating over an array, it is possible to accidentally exceed the bounds of the array. One thing to note about the indices of an array: 

- The first element of an array is stored at index 0. This is true for JavaScript as well as most other programming languages. Therefore, the index of the last element of an array is equal to `arr.length - 1`.

When constructing a `for-loop`, there are a couple ways to handle exceeding the bounds of an array. They both involve altering the conditional statement in the `for-loop` to either:

- `i < arr.length` or
- `i <= arr.length - 1`

```js
const arr = [ "foo", "bar", "baz" ];

// Replace the <= with < in the condition statement
for (let i = 0; i < arr.length; i++) {
	console.log(arr[i]);
}

// Output:
// foo
// bar
// baz

// Replace arr.length with arr.length - 1
for (let i = 0; i <= arr.length - 1; i++) {
	console.log(arr[i]);
}

// Output:
// foo
// bar
// baz
```

## 2.`for … of` loop
```js
for (variable of object) {
	// do something
}
```

> The `for … of` loop can be used to iterate over arrays, sets, and maps. When using the `for … of` loop, there’s no danger of going out of bounds! The loop will go over each item in the iterable object.

Reviewer: Consider adding a callout or including a link to maps. It's not explicitly introduced within the topic and its name is different enough from dictionaries and hashtable that they may not make the connection it's a hashmap.

### Examples

**1. Iterate over items in a list**
```js
const monsters = [
	"Michael Myers", 
	"Jason Voorhees", 
	"Freddy Krueger"
];

> for (let monster of monsters) {

Reviewer: Consider expanding upon the `let  monster of monsters` syntax. This may look unfamiliar because Python doesn't require any keywords to define a new variable. May also be a good opportunity to reinforce previous material on `let` vs `var` and `const`. Why use `let` here?

	console.log(monster);
}

// Output:
// Michael Myers
// Jason Voorhees
// Freddy Krueger
```

**2. Iterate over items in a map**
```js
const m = new Map();
m.set(1, "red");
m.set(2, "black")

for (let item of m) {
	console.log(item);
}

// Output:
// [1, red]
// [2, black]
```

## 3. `for … in` loop
```js
for (property in object) {
	// do stuff
}
```

The `for … in` loop iterates over all enumerable properties of an object. The code in the loop is executed for each property in the object. 

### Example
```js
const movies = {
	2008: "The Dark Knight",
	2009: "Avatar",
	2010: "Toy Story 3"
};

for (let key in movies) {
	console.log(key + ": " + movies[key]);
}

// Ouput:
// 2008: The Dark Knight
// 2009: Avatar
// 2010: Toy Story 3
```

### Common Pitfall: Unexpected Behavior When Iterating over an Array

> While you can use the `for … in` to iterate over an array, it is recommended to use one of the aforementioned for-loops instead. 

Reviewer: Consider using 'we' over 'you' where possible.
```suggestion
While we can use the `for … in` to iterate over an array, it is recommended we use one of the aforementioned for-loops instead. 
```

The reason for this is because the `for … in` loop iterates over **all** of the enumerable properties in an object, including any that may be inherited.

> If, for instance, a JS library modifies the `Array` prototype, the loop will iterate over any additional properties as well. Check out the example below for a potentially non-intuitive demonstration.

Reviewer: I'm not sure they are familiar with what a prototype is. We may need to expand on what the prototype property is. It might be enough to say the reasoning has to do with something called prototypes which are out of scope for this course, and include a link so they can follow their curiosity 💫

```js
const arr = [1, 2, 3];

Array.prototype.newMethod = true;

for (const i in arr) {
	console.log(i);
}

// Output:
// 1
// 2
// 3
// newMethod
```

It is also worth mentioning that the `for … in` loop is meant for objects, and thus will be slower than other loops which are better suited for arrays.

### Practical Usage

So, wait, why use the `for … in` loop at all?

The `for … in` loop may be most practically used as a tool to debug. It is an easy way to check the properties of an object. 

## 4. while loop
```js
while (condition) {
	// do some repetitive task
}
```

The `while` loop will evaluate the `condition` before the loop is run each time. If the `condition` evaluates to `true`, the loop executes the statement(s) in the block. Otherwise, the loop exits and stops executing the statement(s) in the block.

> The `while` loop is known as a pre-test loop because it evaluates the `condition` *before* each iteration. The loop will never execute if the condition evaluates to `false` before entering the loop.

Reviewer: Consider eliminating the mention of pre-test loops. In my view this is extra jargon that they won't come across much. 

```suggestion
`While` loops evaluate the `condition` *before* each iteration. The loop will never execute if the condition evaluates to `false` before entering the loop.
```

### Example

```js
let count = 5;
while (count > 0) {
	console.log(count);
	count -= 1;
}

// Output:
// 5
// 4
// 3
// 2
// 1
```

Try converting the following `for-loop` into a `while` loop:
```js
for (int i = 0; i <= 20; i += 5) {
	console.log(i);
}
```

<details style="max-width: 700px; margin: auto;">
  <summary>
  Click here to reveal a possible solution
  </summary>

  ```javascript
    let i = 0;
    while (i <= 20) {
        console.log(i);
        i += 5;
    }

    // Output:
    // 0
    // 5
    // 10
    // 15
    // 20
  ```
</details>

## 5. do…while loop
```js
do {
	// perform repetitive task at least once
} while(condition)
```

The `do...while` statement creates a loop that executes a block of code until a condition evaluates to `false`.

> The statements in the block will **always** be executed once before the condition is checked. This is what differentiates a `do...while` loop from a `while` loop. The `do...while` loop is sometimes called a post-test loop.

Reviewer: Again, I would suggest eliminating mentions of pre and post test, but I could definitely be persuaded otherwise.

```suggestion
In contrast to a `while` loop, a `do...while` loop will **always** execute the statements in the block once _before_ the condition is checked. 
```

### Example
```js
let i = 1
do {
	console.log(i);
	i += 1;
} while (i <= 5)

// Output:
// 1
// 2
// 3
// 4
// 5
```

### Common Pitfall: Indefinite Loop

> It’s important to ensure the variable(s) used in the statement are updated such that the `while(condition)` evaluates to false. Otherwise, you may crash your program due to an infinite loop.

Reviewer: ```suggestion
It’s important to ensure the variable(s) used in the condition are updated such that the `while(condition)` evaluates to false. Otherwise, our program may crash due to an infinite loop.
```

- `myArray` is the name of the array through which items you would like to iterate.
- `myFunction` is the name of the function you would like to run for each of the elements in the array.
- `currentValue` an element in the array which is currently being iterated upon. It is the only variable which is required to be passed into the function `myFunction`. This is the variable that will represent each element in the array.
- `index` is an **optional** variable used to indicate the index of the current element in the array which is being iterated upon.
- `arr` is an **optional** variable which is the array of the current elements.

### Example
```js
let instructors = ["Auberon", "Char", "Claire", "Kyra"];

function sayHello(instructor) {
	console.log(`Hello, ${instructor}!`);
}

instructors.forEach(sayHello);

// Output:
// Hello, Auberon!
// Hello, Char!
// Hello, Claire!
// Hello, Kyra!

function addHello(instructor, index, arr) {
	arr[index] = 'Hello, ' + instructor;
}

instructors.forEach(addHello);

console.log(instructors)

// Output:
// ['Hello, Auberon', 'Hello, Char', 'Hello, Claire', 'Hello, Kyra']
```

> ### break

Reviewer: Consider adding a paragraph on coding style and `break` and `continue` statements. While they can be useful, we often want to minimize their usage because they can make the execution flow unpredictable/difficult to follow.
